### PR TITLE
apply upcoming lint cast_nullable_to_non_nullable

### PR DESCRIPTION
**Pull request by a14n**
## Description

apply upcoming lint cast_nullable_to_non_nullable

## Related Issues

* dart-lang/sdk#40146

## Tests

I added the following tests:

None

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/table.dart [hunk 1]
```diff
@@ -280,7 +280,7 @@ class _TableElement extends RenderObjectElement {
   Table get widget => super.widget as Table;
 
   @override
-  RenderTable get renderObject => super.renderObject as RenderTable;
+  RenderTable get renderObject => super.renderObject! as RenderTable;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/table.dart:283]: This pattern strikes me as odd. Why is renderObject nullable in the super-class, but not in (many of its) subclasses?

File: packages/flutter_test/lib/src/binding.dart [hunk 2]
```diff
@@ -1466,7 +1466,6 @@ class LiveTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
 
   @override
   void initRenderView() {
-    assert(renderView == null);
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/binding.dart:1469]: Why remove this? Don't we have to keep these asserts for apps that run in mixed mode for now?

File: packages/flutter/lib/src/widgets/table.dart (hunk 1 above)
Comment (a14n) [reply packages/flutter/lib/src/widgets/table.dart]: `RenderObjectElement.renderObject` is never null once mounted.

File: packages/flutter_test/lib/src/binding.dart (hunk 2 above)
Comment (a14n) [reply packages/flutter_test/lib/src/binding.dart:1469]: This is the only place where `renderView` thrown with the change :
```diff
--- a/packages/flutter/lib/src/rendering/binding.dart
+++ b/packages/flutter/lib/src/rendering/binding.dart
@@ -163,7 +163,7 @@ mixin RendererBinding on BindingBase, ServicesBinding, SchedulerBinding, Gesture
   late PipelineOwner _pipelineOwner;
 
   /// The render tree that's attached to the output surface.
-  RenderView get renderView => _pipelineOwner.rootNode as RenderView;
+  RenderView get renderView => _pipelineOwner.rootNode! as RenderView;
   /// Sets the given [RenderView] object (which must not be null), and its tree, to
   /// be the new render tree to display. The previous tree, if any, is detached.
   set renderView(RenderView value) {
```

**Review by Hixie — APPROVED** [review]
LGTM

I didn't check for false negatives (cases where this should be applied but is not).

These changes look great though. Much more readable in terms of being able to see where we're doing "dangerous" casts.

cc @leafpetersen - I think this is a lint we should enable by default.

**Comment by a14n** [comment]
@Hixie or @goderbauer : could you take a look and let me know if I can merge?

> Google testing — Google testing failed. Please contact a Google developer to investigate.

**Comment by Hixie** [comment]
Errors look like this:
```
flutter/lib/src/cupertino/action_sheet.dart:430:30: Error: The return type of the method '_CupertinoAlertRenderElement.renderObject' is '_RenderCupertinoAlert?', which does not match the return type, 'RenderObject', of the overridden method, 'RenderObjectElement.renderObject'.
 - '_RenderCupertinoAlert' is from 'package:flutter/src/cupertino/action_sheet.dart' ('flutter/lib/src/cupertino/action_sheet.dart').
 - 'RenderObject' is from 'package:flutter/src/rendering/object.dart' ('flutter/lib/src/rendering/object.dart').
Change to a subtype of 'RenderObject'.
  _RenderCupertinoAlert? get renderObject => super.renderObject as _RenderCupertinoAlert?;
                             ^
flutter/lib/src/widgets/framework.dart:5436:20: Context: This is the overridden method ('renderObject').
  RenderObject get renderObject => _renderObject;
                   ^


ERROR: flutter/lib/src/cupertino/action_sheet.dart:430
  '_CupertinoAlertRenderElement.renderObject' ('_RenderCupertinoAlert? Function()') isn't a valid override of 'RenderObjectElement.renderObject' ('RenderObject Function()'). #invalid_override
```

**Comment by goderbauer** [comment]
That may be happening because some of the other NNBD PRs have not rolled into google yet.

**Comment by leafpetersen** [comment]
> cc @leafpetersen - I think this is a lint we should enable by default.

cc @natebosch @mit-mit @munificent

**Comment by natebosch** [comment]
> I think this is a lint we should enable by default.

Will we plan on disabling this lint (or even removing it) after Dart 3.0 or do we expect this to be the pattern forever?

I'm _slightly_ hesitant to default enable a lint that is really only useful for unsound mode and wouldn't be how code is written in a sound null safe world.

**Comment by leafpetersen** [comment]
> Will we plan on disabling this lint (or even removing it) after Dart 3.0 or do we expect this to be the pattern forever?
> 
> I'm _slightly_ hesitant to default enable a lint that is really only useful for unsound mode and wouldn't be how code is written in a sound null safe world.

I think this is intended to be used in sound mode code as well.  I believe the argument is that it's easy to do a downcast from (say) `Object?` to a specific type, only thinking about the intended target type but forgetting to consider the nullable case.  This lint encourages using the pattern `x! as int` if you really intend to both discard the nullability information and do the downcast.  Essentially it makes `as` be a cast that works only *within* a specific nullability hierarchy, but not a way to go *between* the two nullability hierarchies, and makes you use `!` to move between the two hierarchies.   I believe this PR includes some fixes to code that was in fact erroneous, is that right @a14n ?  

I don't have strong feelings about this lint, but I think it's worth evaluating whether we wish to enable it in some or all of our code.  

cc @davidmorgan

**Comment by Hixie** [comment]
Yes, I was proposing this to be the pattern forever.

**Comment by a14n** [comment]
> I believe this PR includes some fixes to code that was in fact erroneous, is that right @a14n ?

Yes! Everywhere I added a `?` (`x as T` => `x as T?`) there was an issue.

**Comment by natebosch** [comment]
Ah, I had a misunderstanding about the behavior of the lint.

I do not think we can or should enable this by default for projects during the migration period. We have places where we intentionally use `as` instead of `!` so that unsound code can continue to use `null`, while sound mode code would fail. This allows minor "breaking" changes that only happen when enabling sound mode.

**Comment by a14n** [comment]
>I do not think we can or should enable this by default for projects during the migration period. We have places where we intentionally use as instead of ! so that unsound code can continue to use null, while sound mode code would fail. This allows minor "breaking" changes that only happen when enabling sound mode.

@natebosch I would be very confused if `value as Object` didn't throw if value contains null. It looks like a bug to me. @leafpetersen is it the expected behaviour?

That said, in migrated code if someone use `x = y as Object` then it's really likely the x is used as non null and any method calls will throw. I don't really understand what is the advantage.

**Comment by natebosch** [comment]
> I don't really understand what is the advantage.

The advantage is allowing backwards compatibility in opt-out code while still giving a better API for opt in code.

Perhaps the easiest way to explain is with a concrete example:
The `Iterator<T>.current` getter returns a potentially non-nullable type `T`, but some legacy code might read `current` before calling `moveNext()` and expects to get `null`. We want the signature to be `T`, because needing to use `iterator.current!` in 100% of uses would be annoying, but we implement with `as T` instead of `!` because otherwise it would be a breaking change that's hard to detect statically. In one of the few places where we missed this and used `!` instead it was quickly noticed by a user. https://github.com/dart-lang/sdk/issues/42950


> I would be very confused if `value as Object` didn't throw if value contains null. It looks like a bug to me.

Yes, this is expected behavior. In unsound mode `as` continues to allow `null` as a value, while `!` is a null check in both unsound and sound mode.

**Comment by a14n** [comment]
Thanks for the explanation. Even if I understand the reasoning it looks really strange to see `value as String` and imagine value can even be null.

**Pull request merged**